[5.5] API Responses

themsaid opened the pull request:
This PR Introduces a trait you can use on a `Responsable`:

```php
class UserApiResponse implements Responsable
{
    use isApiResponse;

    public function __construct($resource)
    {
        $this->resource = $resource;
    }

    public function transformResource($resource)
    {
        return [
            'name' => $resource->name,
            'email' => $resource->email,
        ];
    }
}
```

The trait needs a `$resource` property to exist on your Responsable object, using this convention it's going to create a response for you based on the type of object you pass (Resource, Collection, Paginator), you have to implement a `transformResource` method to tell Laravel how to transform a single resource instance for the API.

Now in your controller you can return any of the following:

```php
return new UserApiResponse(User::find(1));

return new UserApiResponse(User::paginate(15));

return new UserApiResponse(User::find([1,2]));

// You can pass a status code using:
return (new UserApiResponse($user))->withStatus(201);

// You can also do:
return UserApiResponse::with($user)->withStatus(201)->withMeta(['valid' => true]);
```

The responses will look like the following:

For a single resource:

```
{
  "data": {
    "name": "Prof. Alfonzo Schulist DDS",
    "email": "helen.dickinson@example.com"
  },
  "meta": {
    "value": true
  },
}
```

For when you pass a paginator:

```
{
    "data": [
        {
            "name": "Prof. Alfonzo Schulist DDS",
            "email": "helen.dickinson@example.com"
        },
        {
            "name": "Prof. Oma Armstrong PhD",
            "email": "iconn@example.org"
        }
    ],
    "meta": {
       "value": true,
        "current_page": 1,
        "from": 1,
        "last_page": 50,
        "next_page_url": "http://laravelnext.dev/controller?page=2",
        "path": "http://laravelnext.dev/controller",
        "per_page": 2,
        "prev_page_url": null,
        "to": 2,
        "total": 100
    }
}
```

For when you pass a collection:

```
{
    "data": [
        {
            "name": "Prof. Alfonzo Schulist DDS",
            "email": "helen.dickinson@example.com"
        },
        {
            "name": "Prof. Oma Armstrong PhD",
            "email": "iconn@example.org"
        }
    ],
    "meta": {
        "value": true
     },
}
```

themsaid committed:
api responsables

arcanedev-maroc commented:
Hi @themsaid, can you change the `metadata` name to `meta` ?

Keep up the good work 👍

arcanedev-maroc commented:
Hi again,

can you check the paginator response in your description:

```json
{
    "data": [
        {
            "name": "Prof. Alfonzo Schulist DDS",
            "email": "helen.dickinson@example.com"
        },
        {
            "name": "Prof. Oma Armstrong PhD",
            "email": "iconn@example.org"
        }
    ],
    "meta": {
        "current_page": 1,
        "from": 1,
        "last_page": 50,
        "next_page_url": "http://laravelnext.dev/controller?page=2",
        "path": "http://laravelnext.dev/controller",
        "per_page": 2,
        "prev_page_url": null,
        "to": 2,
        "total": 100
    }
}
```

The `last_page` must be equals to `2` and the `per_page` equals to `50` (also the `to` value).

yajra commented on src/Illuminate/Http/ApiResponseTrait.php:
```diff
@@ -0,0 +1,35 @@
+    {
+        if ($object instanceof Collection) {
+            return [
+                'data' => $this->user->map([$this, 'transformResource']),
```
I think this should be `$object` and not `$this->user`? Same applies below.

sixlive commented:
Would love to see a static factory especially with the `withStatus()` method.

```php
return UserApiResponse::create(User::find(1))->withStatus(201);
```

I would also get some usage about adding misc meta. So maybe a `->withMeta($metaArray)` or `->withMeta($key, $value)` that would be merged in with the patination data if that exists.

Any thoughts on making the pagination meta its own key?

```json
{
  "data": [
    {
      "name": "Prof. Alfonzo Schulist DDS",
      "email": "helen.dickinson@example.com"
    },
    {
      "name": "Prof. Oma Armstrong PhD",
      "email": "iconn@example.org"
    }
  ],
  "meta": {
    "pagination": {
      "current_page": 1,
      "from": 1,
      "last_page": 50,
      "next_page_url": "http://laravelnext.dev/controller?page=2",
      "path": "http://laravelnext.dev/controller",
      "per_page": 2,
      "prev_page_url": null,
      "to": 2,
      "total": 100
    }
  }
}
```

themsaid commented:
Added a `::with()` static constructor as well as a `withMeta()` method.

```
return UserApiResponse::with(User::paginate(5))->withMeta(['test' => 123]);
```

themsaid committed:
code style

JacobBennett commented:
This looks like an awesome addition! Would help a ton to be able to standardize the format that we use to return api data to the front end. With that in mind, are there any thoughts regarding a standard format when returning errors for api requests? For instance if I use a `User::findOrFail(1)`, currently that will generate a HTML response instead of a `JSON` response. Maybe if using a `Responsable`, we can assume they want a `JSON` error instead and handle it appropriately? 

Not sure if this is the correct venue for discussing it but figured I would give it a mention.

sixlive commented:
@JacobBennett I typically do this in the global exception handler since most of those types of exceptions you will want to handle in the same way. Also, you will want to look at the way 5.5 is going to render exceptions now, you might be able to bake it into that.

themsaid commented:
@JacobBennett in 5.5 if your API call has a `Accept: application/json` header laravel will transform all exceptions to JSON:

```
{
    "message": "Exception message"
}
```

sixlive commented:
Totally forgot to mention that. Would be nice if it would respond with JSON automatically if the route is an API route

arcanedev-maroc commented:
@sixlive, in your API routes (or Controllers) should respond with:

```php
return response()->json(...);
```

Or something similar.

This PR is just a helper to transform objects without using external packages like fractal or even using `response()->json(...)`.

sixlive commented:
Thanks! Well aware of what this PR is. Just making a side comment on global exception handling responses.

DanielDarrenJones commented:
@sixlive This was added in a recent PR to fix for 5.5 #18732 and has been merged, the API should be standardised across Laravel going forward hopefully 👍

sixlive commented:
@DanielDarrenJones Oh sweet! Thanks mate! Totally missed that one!

lkmadushan commented:
Isn't it better to use class inheritance without using a trait. Something like,

```
ApiTransformer implements Responsable {
     ..
}
UserTransformer extends ApiTransformer {
    ...
}
```

Anyway it seems this is equivalent to league/fractal

JacobBennett commented:
Thanks for the heads up on the JSON response stuff for exceptions @themsaid, looks like 5.5 has it covered!

a-komarev commented:
@lkmadushan you always could make your own class which will use trait under the hood.

tillkruss commented on src/Illuminate/Http/ApiResponseTrait.php:
```diff
@@ -0,0 +1,107 @@
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Pagination\AbstractPaginator;
+
+trait ApiResponseTrait
```
Could we call this something else? Like `ApiResponse` or `RespondsToApiCall` or something without the `Trait` suffix..

a-komarev replied:
Agree with @tillkruss. Suggesting one more option: `HasApiResponse`

JosephSilber replied:
This trait actually describes what the class _is_, so `IsApiResponse` sounds good to me.

a-komarev replied:
@JosephSilber option is the best one.

lucasmichot commented on src/Illuminate/Http/ApiResponseTrait.php:
```diff
@@ -0,0 +1,107 @@
+
+use InvalidArgumentException;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Pagination\AbstractPaginator;
```
Then we should also require `"illuminate/pagination": "5.5.*"` in `Illuminate/Http/composer.json`

lucasmichot commented on src/Illuminate/Http/ApiResponseTrait.php:
```diff
@@ -0,0 +1,107 @@
+    /**
+     * Create an HTTP response that represents the object.
+     *
+     * @return \Illuminate\Http\Response
```
`@return \Illuminate\Http\JsonResponse` instead

lucasmichot commented on src/Illuminate/Http/ApiResponseTrait.php:
```diff
@@ -0,0 +1,107 @@
+     * Create an HTTP response that represents the object.
+     *
+     * @return \Illuminate\Http\Response
+     * @throws InvalidArgumentException
```
Missing backslash => `@throws \InvalidArgumentException`

lucasmichot commented:
Great @themsaid !

I think having `withHeaders()` and `withOptions` would be great too, so that we can use the headers and options values when returning the response:
`return response()->json($this->transform(), $this->status, $this->headers, $this->options);`

a-komarev commented on src/Illuminate/Http/ApiResponseTrait.php:
```diff
@@ -0,0 +1,107 @@
+     *
+     * @var array
+     */
+    public $metaData = [];
```
Can't agree with it because result will be displayed in `meta` array key.

themsaid commented:
Thanks @lucasmichot and @JosephSilber, applied the following changes:

- Renamed to `isApiResponse`
- Renamed `metaData` to `metadata`
- Added `illuminate/pagination` to composer.json
- Changed the import to look for `support/collection` instead of `eloquent/collection`
- Added `withHeaders`
- Added `withEncodingOptions`

lucasmichot commented on src/Illuminate/Http/composer.json:
```diff
@@ -17,6 +17,7 @@
         "php": ">=7.0",
         "illuminate/session": "5.5.*",
         "illuminate/support": "5.5.*",
+        "illuminate/pagination": "5.5.*",
```
Should probably be sorted ;-)

themsaid committed:
multiple changes

themsaid committed:
update tests

JosephSilber commented:
Shouldn't we use the `ResponseTrait` here?

Aferz commented:
This looks pretty similar to my PR #18502 ...

deleugpn commented on src/Illuminate/Http/IsApiResponse.php:
```diff
@@ -0,0 +1,149 @@
+    public function toResponse()
+    {
+        if (isset($this->resource)) {
+            return response()->json($this->transform(),
```
Since it's responding with `->json()`, wouldn't it make sense to call the trait `JsonResponse`?

a-komarev replied:
@deleugpn What if developer overwrite `toResponse` method in parent class and it will return XML? I like current trait name.

lucasmichot commented on src/Illuminate/Http/IsApiResponse.php:
```diff
@@ -0,0 +1,149 @@
+                    : $this->transformResource($object);
+
+        if ($object instanceof AbstractPaginator) {
+            $meta = array_merge($meta, array_except($object->toArray(), ['data']));
```
`\Illuminate\Support\Arr::except` maybe ?

taylorotwell commented:
I think if we are going to adopt some kind of standard JSON response I would rather stick to an established spec like JSON-API, etc. It would give us less to argue about and would be easier to learn for others.

a-komarev commented:
JSON API seems to be the most common this days and has most active community if compare to other standards.

adiachenko commented:
@taylorotwell But JSON-API is way outside the scope of standard API response structure. It also covers requests, content negotiation, query filtering, sparse fieldsets and a lot more. 

We're continuously postponing simple quality of life changes on the premise that Laravel will have something much more robust years away from now.

Just looking at [existing implementations](http://jsonapi.org/implementations/#server-libraries), none of them have the level popularity and spec compliance that inspire any confidence.

lucasmichot commented:
I share @adiachenko opinions on that, @taylorotwell 
@themsaid, any chance to release this code as a package ?

themsaid commented:
@taylorotwell said he'll think about it more, if he decided to not go for it then yes I think releasing it as a package would be nice :) Just hope this gets into the core since it's really simple but helps with creating responses a lot.

DanielDarrenJones commented:
I fully agree, JSON-API feels way to over spec for what 99% of people need.

Having something simple now would make such a huge improvement across the board for Laravel, is there a way the formatters could be pulled out to be configurable so if in the future it changes to JSON-API or another mapping it can be replaced by the user?

joselfonseca commented:
This looks like what http://fractal.thephpleague.com does.

deleugpn commented:
@joselfonseca I would go out on a limb here and say that 100% of people trying to PR API responses into Laravel have used and disliked Fractal.

DanielDarrenJones commented:
@deleugpn Actually I really like fractal, especially when wrapped in https://github.com/spatie/laravel-fractal, that said it would be nice to have a unified global api response system in Laravel.

joselfonseca commented:
@deleugpn Interesting, I find it quite good myself. Then again having something in Core would be nicer.

adiachenko commented:
@DanielDarrenJones Well, Fractal (kinda) does the job, but it's very awkward.  Among other things:

- it causes severe N+1 problems unless you are very careful about what you're doing and explicitly check whether relationship is loaded in Transformer's includes (which isn't what suggested in the docs)
- it performs [complex query string parsing](https://github.com/thephpleague/fractal/blob/0.16.0/src/Manager.php#L158) on each request
- apparently, [it has very rigid serializers](https://github.com/thephpleague/fractal/blob/0.16.0/src/Serializer/DataArraySerializer.php#L14). Say you namespace you response under `data` property, this will automatically propagate to relationships based on how Serialize interface is built which may not be what you want
- it claims to offer multiple serializers out of the box, but it is apparently built with `DataArraySerializer` in mind and everything else is just an extra (the one for JSON API won't work in a real-world usage and I don't even know what to say about HAL serializer that is mentioned in the docs but is simply not there)
- etc.

In short, there's no smoke without fire. Otherwise, we'll all be using Fractal by now instead of trying to push a similar solution into the core.

themsaid committed:
fix conflicts

arcanedev-maroc commented:
Praise the 🌞 \[T]/

taylorotwell commented:
Something that would help me in this thread are answers to this question: what are we trying to solve? At the end of the day, implementations seems to boil down to something like this:

```
function transformPost($post)
{
    return [...];
}

function transformUser($user)
{
    return [...];
}
```

Everytime I start hacking on it a bit I come back to the core fact that we are just giving you a fancy way to declare a plain PHP function that returns an array. Is that all we're trying to do? Does everyone agree with that?

themsaid commented:
@taylorotwell basically yes, a way to convert an object to a PHP array to finally be converted to JSON.

We're also trying to bring the ability to automatically convert a Collection of objects or a pagination to a JSON format, where objects inside those collections are transformed using the toArray function we're adding in this feature.

taylorotwell commented:
And the main problem with using `toArray` on the model to do this is because you have multiple API versions? Or is there some other reason?

DanielDarrenJones commented:
@taylorotwell 

The main reason for transformers in my opinion is to decouple the underlying attribute names or database table names from the API response, doing do means we can change the underlying data structure while still keeping consistencies for users.

It also gives us a chance to cast the type explicitly before it is transformed to JSON, this has its uses from time to time.

Having a base transformation structure of the model allows us to pull that in a different response templates and transform to standard API response formats much like JSONAPI or similar.

I think ultimately I am looking for something very similar to https://github.com/spatie/laravel-fractal, I want an easy way to model my API responses. 

I think over the last year and a half especially since Laravel has become so close to Vue we have seen a strong shift towards API driven development and for Laravel to fully embrace this we need the tools to do it correctly.

arcanedev-maroc commented:
I think the *main* reason of this feature is to have something similar to `thephpleague/fractal` out of the box.

Creating a plain PHP function doesn't seems as an *Eloquent way* to work with.

When creating a **Transformer** class, you can apply it to a `Model`, `Eloquent Collection` or a `Pagination` object.

Instead of using an external library, i think laravel already has more than `90%` of features to handle the json responses (especially the `transform` method in `Collection` class). 

Adding this feature is like *adding the cherry on the cake*.

themsaid commented:
@taylorotwell `toArray` could be used it tests or other places in the code, it's useful to have a function that's only purpose is to convert a resource to an API format, maybe do conversions, change float to integer, append strings, hide parts of a secret `*******9876`, etc...

taylorotwell commented:
OK thanks for your responses.

lkmadushan commented:
Is there any benefit using this instead of league/fractal?

taylorotwell commented:
@lkmadushan someone tried to answer that like 6 comments up.

lkmadushan commented:
@taylorotwell yeah. see those comments. but didn't see any special difference compared to league/fractal. it seems like we are doing same thing in different ways.

but It would be great if you can provide a JSON-API response format and a supported client for the frontend to consume JSON-API API service easily.

Aferz commented:
For me, the main purpose of features like this is being able to respond to a request in a very specific way with very specific data with *common* resources.

With attributes like `hidden`, `append`, or `with` inside models, we are just being able to declare *general* behaviour to this model for the whole application. Despite this is not a _big_ problem, with this configuration we are assuming that when we want to display an user, we always want to display them with the same attributes, and we all know this is not true for every scenario.

We are trying to use common resources (Models) for specific responses. That's why generic behaviours doesn't match very well when we are trying to do such specific/independent things.

Fractal does a very good work in this area, but it is extremely verbose for a framework like Laravel and I and many like me think it could be done better. (No pretending to offense to PHPLeague, of course)

That's why I opened a PR (time ago) trying to figure out what is the Laravelish way to accomplish such work #18502. Is a bit old, and I didn't know about recent `Responsable` interface, so it could be handled better now, but you get the basic idea.

deleugpn commented:
@lkmadushan it's the same thing, but done right.

taylorotwell commented:
@deleugpn I'm curious to hear your problems with Fractal.

a-komarev commented:
@lkmadushan JSON:API Server responses is a good thing to have out of the box, but JSON:API Client is task out of the Laravel scope. It's not a trivial job and there is no reason for the Laravel team to handle it.

deleugpn commented:
@taylorotwell at first, it's not beginners-friendly. You have to opt for `dingo/api` or `spatie/laravel-factral`. Although spatie does provide a good support for it, it's still pretty verbose.

As you mentioned yourself in the thread, we're basically seeking model-to-array transformation layer. The amount of overhead just for that is too high through Factral.

Factral `include` functionality is actually a Pandora's box. You think it's making your life easy by offering relationship transformation, but you end up with (already mentioned) N+1 problem unless you're extremely careful with covering all your entry points to the layer.

At the end of the day, I think we're trying to have simple (and laravel-friendly) transformers for simple things.

taylorotwell commented:
Done.

arcanedev-maroc commented:
**PR:** #20710
